Add current working directory [:default-directory] property

sergv opened the pull request:
Hi, this is a followup to #813. I cannot add new commits to that pull request so I'm creating a new one.

Since original pull request was stalled for quite some time with several unresolved review comments I decided to jump in and resolve any mentioned issues. This feature is pretty essential for Haskell checking, in my opinion.

Stylistic issues were resolved and, I hope, no new were introduced.

The problem with `flycheck-fill-and-expand-error-file-names` not having access to current working directory was fixed as well. The issue was critical for checkers that can report relative file names in their errors, but it didn't affect checkers that report absolute file names. So it was not as critical as last comment of #813 suggests, but it's fixed now anyway.

To fix it, I added `default-directory` field to `flycheck-syntax-check` structure  and moved binding of `default-directory` variable out of `flycheck-start-command-checker` function into the `flycheck-syntax-check-start` function. This indirection obscures things a bit but allows to calculate default directory only once per cherker run. Since computing default directory may involve costly traversals of filesystem, it seems worth some additional obscurity.

I regret not adding any new test cases, but it seems to require some non-trivial effort - the core repository seems to contain only, well, core tests and no tests for specific checkers.

Edit (@lunaryorn): Dear Waffle, this connects to #312.

lunaryorn commented:
@sergv Wow, thanks for taking up :+1:  Really awesome work; at a cursory look I really like how you did things here.  Also, many many thinks for fixing all those stylistic issues I left behind ☺️ 

I'll review and test this PR soon, but please be patient.  I'm a little short of time this week, and likely won't get to review this PR until next week.

sergv commented:
I feel uneasy for hijacking this pull request with fix for #978, but the fix is slightly dependent on changes that introduced `cwd` argument in the `flycheck-fix-error-filename` function.

lunaryorn commented:
@sergv Please remove that change from this PR.  It's unrelated, and while I'm all in favour of a `default-directory`, I'm not convinced that replacing file names in error messages is a good idea.

Please let's discuss this separately, even if this means that the fix for #978 will be delayed until after this PR is merged.

sergv commented:
Okay, I removed the fix for #978 from here, it can wait until this PR is merged.

lunaryorn commented on test/flycheck-test.el:
```diff
@@ -1665,13 +1665,18 @@
+                                                                        cwd))
                      (list (flycheck-ert-resource-filename "foo")
                            absolute-fn
+                           absolute-fn
```
This looks like a spurious whitespace change?

sergv replied:
Good catch but actually it isn't - this inline list gets compared against `errors` list, which has 4 entries now.

lunaryorn replied:
Oh dear, I should have taken a better look.  Sorry for the noise :blush:

lunaryorn commented on flycheck.el:
```diff
@@ -4263,6 +4276,24 @@ universal prefix arg, and only the id with normal prefix arg."
+not a relative one."
+  (let* ((cwd (flycheck-checker-get checker 'default-directory))
+         (default-directory (or (and (functionp cwd) (funcall cwd))
+                                (and (stringp cwd) (file-truename cwd))
```
I don't think that this feature has any use.  Why would you ever want to give a constant string?

I think we can safely remove this part.

lunaryorn commented:
@sergv Can you please also remove b281f2c from this PR?  While it's a good intention it clutters the diff and makes it harder to review the important parts of this PR.

----

I've taken some time to think about the general structure of this PR, and I think that it makes the default directory too, well, "global".  

I don't think that it should be a property of generic syntax checkers—it only makes sense for command syntax checkers in my opinion.  Hence I don't think it should be part of the syntax check structure as well.

Instead, I'd prefer if `default-directory` was only part of `flycheck-define-command-checker`, and stored in the process object returned by this function.  I think it'd be easy enough to attach the working directory  with `(process-put process 'flycheck-default-directory default-directory)`, and later use `process-get` to extract it.

Could you try to refactor the PR accordingly?

sergv commented:
@lunaryorn I was under the impression that unresolved stylistic problems of #813 were, at least in part, blocking the review of the PR. I can remove the fixes and make another PR for them or I can move the commit to be the last commit, so you'll be able to review meaningful part at the beginning without clutter. Which way would be the best?

I'll try refactoring PR and move default-directory into flycheck-define-command-checker soon.

lunaryorn commented:
@sergv Sure, but there are many stylistics changes in this PR that are not related to the actual diff.  For instance, you added periods to a couple of comments that weren't touched by this PR in any other way.  That's not a bad thing per se, but it's cluttering the PR a bit.

It'd be great of you could limit the diff to those parts that are actually changed as part of implementing the default-directory property.  If it's too hard though leave the stylistic fixes in; it's ok.

Thanks for your great work on this PR!

sergv commented:
@lunaryorn I rebased all the changes on latest master. I also reordered commits so that stylistic fixes are at the top and can be easily ignored via `git diff 99c26f6..cwd^`, for example. In addition, I moved default-directory setting back into checker process object and disallowed non-function values of :default-directory setting3.

lunaryorn commented on flycheck.el:
```diff
@@ -4257,6 +4263,27 @@ universal prefix arg, and only the id with normal prefix arg."
     (`(eval ,_) t)
     (_ nil)))
 
+(defun flycheck-default-directory-wrapper (checker)
```
I don't like the name of this function :)  It's too generic.

Can we come up with a better one?  E.g. `flycheck-compute-default-directory`, or so?

lunaryorn commented on flycheck.el:
```diff
@@ -4257,6 +4263,27 @@ universal prefix arg, and only the id with normal prefix arg."
+
+Work out the working directory from which the CHECKER command
+will be called.  The default is the `default-directory' of the source-buffer
+from which flycheck is calling the checker.
```
Wrapping looks off here.  Please rewrap the docstring.

lunaryorn commented on flycheck.el:
```diff
@@ -4257,6 +4263,27 @@ universal prefix arg, and only the id with normal prefix arg."
+(defun flycheck-default-directory-wrapper (checker)
+  "Function returning apropriate current working directory.
+
+Work out the working directory from which the CHECKER command
```
Maybe "Compute" rather than "work out"?

lunaryorn commented on flycheck.el:
```diff
@@ -4257,6 +4263,27 @@ universal prefix arg, and only the id with normal prefix arg."
     (_ nil)))
 
+(defun flycheck-default-directory-wrapper (checker)
+  "Function returning apropriate current working directory.
```
Please use imperative style for the first line of a docstring.  Take a look at other docstrings to see the style that we use.

lunaryorn commented on flycheck.el:
```diff
@@ -4257,6 +4263,27 @@ universal prefix arg, and only the id with normal prefix arg."
+         (default-directory (if cwd
+                                (if (functionp cwd)
+                                    (funcall cwd)
+                                  (error "Syntax Checker %S has defined :default-directory which is not a function: %s"
```
Lowercase "checker" please.

lunaryorn replied:
Even better shorten the error message ":default-directory %S of syntax checker %S is not a function".

lunaryorn commented on flycheck.el:
```diff
@@ -4257,6 +4263,27 @@ universal prefix arg, and only the id with normal prefix arg."
+                                         default-directory))
+                              default-directory)))
+    (unless (file-exists-p default-directory)
+      (error "Syntax Checker %S has defined non existant :default-directory %s"
```
Lowercase "checker" please.  Also "non existant" should be "non-existent".

lunaryorn replied:
Maybe it's better to rephrase the whole error message.  How about ":default directory %S of syntax checker %S does not exist"?

lunaryorn commented:
@sergv There are some minor stylistic issues left, but other than that I think it's ready to be merged.

Please address the remaining issues and rebase on `master`.

lunaryorn commented:
And finally, please add information about this change to `CHANGES.rst` and feel free to add yourself as contributor to `doc/people.rst` 😃

sergv commented:
@lunaryorn Wow, that's a lot of details. I fixed them, please review `Tidy docstrings up` commit.

lunaryorn commented:
LGTM.  @sergv Feel free to merge :relaxed:

sergv commented:
@lunaryorn Thanks! Unfortunately, github says that I'm "not authorized to merge this pull request" and the "Merge pull request" button is grayed out :(

lunaryorn commented:
@sergv Can you show me a screenshot?  That's really strange, you should have access to the repo imho.  Maybe I screwed up Github permissions here.

lunaryorn commented:
@sergv Can you try again?  I changed a setting for protected branches; I think I misunderstood how protected branches work.

sergv commented:
@lunaryorn now the button is green, thanks!

sergv commented:
@lunaryorn However, it seems new approval is needed after rebase. Can you please approve again?

sergv commented:
For some reason, github still created a merge commit

mrkkrp commented:
Now that we have this merged, we can make `ghc` and `ghc-stack` be called as if from root directory of project, right? Is there any plans to do this change, or if you point where to look how this new feature works I can open a PR myself.

lunaryorn commented:
@sergv Merge commits are fine for larger branches such as this 😊

@mrkkrp This branch adds a new default-directory property to syntax checker definitions.  Refer to the docstrings for details.

I don't plan to update any syntax checkers following this change but if you think it's reasonable to change some syntax checkers please go ahead an open a pull request.

lunaryorn commented:
@sergv I've changed the interface of the `:default-directory` function.  It now takes the syntax checker symbol as argument, to allow to reuse the same function for multiple checkers while optionally dispatching on the syntax checker, and it may return `nil` now to fall back to the normal default directory, for instance in case a project file isn't found.  I think these changes are reasonable, do you agree?

Meanwhile, would you please take a look at #1005?  This change seems to have broken generic syntax checkers.

sergv commented:
@lunaryorn The change to `:default-directory` looks reasonable. I'll look at #1005.

@lunaryorn @mrkkrp Regarding default-directory property for `haskell-stack-ghc` checker, I can submit one. However this would probably mean that we should add `:default-directory` argument to `flycheck-define-checker` macro since `haskell-stack-ghc` is defined using it. Would you agree with such addition, @lunaryorn?

lunaryorn commented:
@sergv Yes the macro should support all properties of flycheck-define-generic-checker.  It'd be great if you could make this change as well.
